src/core/qgsmaprenderercustompainterjob.cpp

```diff
@@ -329,9 +329,15 @@ void QgsMapRendererJob::drawLabeling( const QgsMapSettings &settings, QgsRenderC
   // Reset the composition mode before rendering the labels
   painter->setCompositionMode( QPainter::CompositionMode_SourceOver );
 
+  std::unique_ptr< QgsFeatureFilterProvider > featureFilterProvider;
+  if ( renderContext.featureFilterProvider() ) {
+    featureFilterProvider.reset( renderContext.featureFilterProvider()->clone() );
+  }
+
   // TODO: this is not ideal - we could override rendering stopped flag that has been set in meanwhile
   renderContext = QgsRenderContext::fromMapSettings( settings );
   renderContext.setPainter( painter );
+  renderContext.setFeatureFilterProvider( featureFilterProvider.get() );
```

Comment:
The `featureFilterProvider` lives only in this method scope, but is set on the `renderContext` which lives outside this method.
Either pass ownership to the `renderContext` or reset the `featureFilterProvider` to its previous value before this method returns.

Reply:
As far as I understand in renderContext keep a clone of the passed featureFilterProvider, see https://github.com/qgis/QGIS/blob/055a166136f29972571944bc741fbc68af42b4ea/src/core/qgsrendercontext.cpp#L244
This is why I want to delete temporary clone.

Reply:
I see. So that's basically a double-clone?

Why not shorten it to

```
QgsFeatureFilterProvider *provider = renderContext.featureFilterProvider();
...
renderContext.setFeatureFilterProvider( featureFilterProvider );
```

Reply:
This result in a SEGFAULT.
It seems that filter provider is destroyed after: 
`renderContext = QgsRenderContext::fromMapSettings( settings );`

Note that QgsRenderContext is a unique_ptr: https://github.com/qgis/QGIS/blob/a7c3aadf3d215774ee1202127718bdaa6437f71f/src/core/qgsrendercontext.h#L543

Reply:
I'm still lost about the purpose of the renderContext in here. The existing context passed in is completely overwritten (which is probably not thread safe, as also indicated by the comment above `::fromMapSettings`) and this commit is the first one to introduce code that preserves anything from the previous renderContext.
I'm probably unaware of some background knowledge that led to this design decision, @wonder-sk do you remember the rationale for replacing the existing context? Should the featureFilterProvider be part of mapSettings?

Reply:
If there are no further insights from @wonder-sk available, I agree it's best to stick to the proposed solution as anything else would require a more intense refactoring.

Reply:
Hmm after looking into the code and history a bit, the whole reset of the render context in drawLabeling() seems useless. From what I understand, in older version of the code, this was the place where the actual render context for labeling got initialized. But since 33eb4bc0 from @nyalldawson there is a new prepareLabelingJob() method which does the same thing earlier (make context from map settings, set painter, extent, labeling engine). So if I didn't miss anything, the drawLabeling() body could be reduced to merely calling `labelingEngine2->run( renderContext );` and thus we would avoid this re-assignment of feature filter there (which indeed feels wrong)